sources/web/datalab/polymer/modules/gapi-manager/gapi-manager.ts

```diff
@@ -414,6 +428,74 @@ class ServerAuth implements GapiAuth {
     return Promise.resolve(true);
   }
 
+  /**
+   * Gets the expiration time for the access token, or null if no token.
+   */
+  private async getAccessTokenExpiry(): Promise<Date | null> {
+    const accessTokenBase64 = Utils.readCookie('DATALAB_ACCESS_TOKEN');
+    if (!accessTokenBase64) {
+      return null;
+    }
+    const json = atob(accessTokenBase64);
+    const accessToken = JSON.parse(json);
+    return new Date(accessToken.expiry);
+  }
+
+  /**
+   * Sends a request to the service to refresh our access token.
+   */
+  private async refreshToken(): Promise<void> {
+    const xhrOptions: XhrOptions = {
+      failureCodes: [401],
+      successCodes: [200],
+    };
+    const refreshPath = '/_auth/refresh';
```

Comment:
I'd prefer we add these to the `ServiceId` enum instead of using the string. We kind of missed them in the last PR when refactoring `GapiManager` too.

Reply:
At the moment, I think of this as an internal implementation detail of the server-side auth code, so it seems cleaner to keep it inside ServerAuth. But I can see an argument for putting it in ServiceId.